evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs

```diff
@@ -0,0 +1,467 @@
+impl<'a, B: Backend> SputnikExecutor<CheatcodeStackState<'a, B>> for CheatcodeStackExecutor<'a, B> {
+    fn config(&self) -> &Config {
+        self.handler.config()
+    }
+
+    fn state(&self) -> &CheatcodeStackState<'a, B> {
+        self.handler.state()
+    }
+
+    fn state_mut(&mut self) -> &mut CheatcodeStackState<'a, B> {
+        self.handler.state_mut()
+    }
+
+    fn gas_left(&self) -> U256 {
+        // NB: We do this to avoid `function cannot return without recursing`
+        U256::from(self.state().metadata().gasometer().gas())
+    }
+
+    fn transact_call(
+        &mut self,
+        caller: H160,
+        address: H160,
+        value: U256,
+        data: Vec<u8>,
+        gas_limit: u64,
+        access_list: Vec<(H160, Vec<H256>)>,
+    ) -> (ExitReason, Vec<u8>) {
+        let transaction_cost = gasometer::call_transaction_cost(&data, &access_list);
+        match self.state_mut().metadata_mut().gasometer_mut().record_transaction(transaction_cost) {
+            Ok(()) => (),
+            Err(e) => return (e.into(), Vec::new()),
+        }
+
+        // Initialize initial addresses for EIP-2929
+        if self.config().increase_state_access_gas {
+            let addresses = self
+                .handler
+                .precompile()
+                .clone()
+                .into_keys()
+                .into_iter()
+                .chain(core::iter::once(caller))
+                .chain(core::iter::once(address));
+            self.state_mut().metadata_mut().access_addresses(addresses);
+
+            self.handler.initialize_with_access_list(access_list);
+        }
+
+        self.state_mut().inc_nonce(caller);
+
+        let context = Context { caller, address, apparent_value: value };
+
+        match self.call_inner(
+            address,
+            Some(Transfer { source: caller, target: address, value }),
+            data,
+            Some(gas_limit),
+            false,
+            false,
+            false,
+            context,
+        ) {
+            Capture::Exit((s, v)) => (s, v),
+            Capture::Trap(_) => unreachable!(),
+        }
+    }
+}
+
+pub type CheatcodeStackState<'a, B> = MemoryStackStateOwned<'a, CheatcodeBackend<B>>;
+
+pub type CheatcodeStackExecutor<'a, B> =
+    CheatcodeHandler<StackExecutor<'a, CheatcodeStackState<'a, B>>>;
+
+impl<'a, B: Backend> CheatcodeStackExecutor<'a, B> {
+    /// Decodes the provided calldata as a
+    fn apply_cheatcode(&mut self, input: Vec<u8>) -> Capture<(ExitReason, Vec<u8>), Infallible> {
+        // Get a mutable ref to the state so we can apply the cheats
+        let state = self.state_mut();
+
+        if let Ok(timestamp) = HEVM.decode::<U256, _>("warp", &input) {
+            state.backend.cheats.block_timestamp = Some(timestamp);
+        }
+
+        if let Ok(block_number) = HEVM.decode::<U256, _>("roll", &input) {
+            state.backend.cheats.block_number = Some(block_number);
+        }
```

Comment:
This will probably be a very long elseif chain,
I wonder if it would be worthwhile to make this customizable, something like:

```rust
struct CheatcodeHandler<H, I = HelmCheatcodes> {
    handler: H,
    interceptor: I
}

trait Interceptor {
   type State // CheatcodeStackState for `HelmInterceptor` ;
   fn intercept(&self, address: Address) -> bool;
   fn apply(&mut self, input: Vec<u8>, state: &mut Self::State) -> Capture<(ExitReason, Vec<u8>), Infallible>;
}
```
but perhaps this would result in a rust borrow checker issue when calling `self.interceptor(input,self.state_mut())`